File: src/protocol/x11/request.cpp
#include "protocol/x11/request.hpp"

#include "core/checked_math.hpp"
#include "protocol/x11/byte_cursor.hpp"

#include <algorithm>
#include <stdexcept>

namespace gw::protocol::x11 {

RequestFramer::RequestFramer(const ByteOrder order,
                             const std::uint16_t maximum_length_units)
    : order_(order), maximum_length_units_(maximum_length_units) {
  if (maximum_length_units_ == 0) {
    throw std::invalid_argument("X11 request length limit must be nonzero");
  }
  request_.bytes.reserve(kCoreRequestHeaderSize);
}

RequestFrameResult
RequestFramer::consume(const std::span<const std::uint8_t> input) {
  if (status_ != RequestFrameStatus::NeedMore) {
    return {status_, 0};
  }

  std::size_t consumed = 0;
  while (consumed < input.size() && status_ == RequestFrameStatus::NeedMore) {
    const auto needed = expected_size_ - request_.bytes.size();
    const auto copy_size = std::min(needed, input.size() - consumed);
    request_.bytes.insert(
        request_.bytes.end(),
        input.begin() + static_cast<std::ptrdiff_t>(consumed),
        input.begin() + static_cast<std::ptrdiff_t>(consumed + copy_size));
    consumed += copy_size;

    if (request_.bytes.size() == kCoreRequestHeaderSize &&
        expected_size_ == kCoreRequestHeaderSize) {
      status_ = inspect_header();
    }
    if (status_ == RequestFrameStatus::NeedMore &&
        request_.bytes.size() == kBigRequestHeaderSize &&
        expected_size_ == kBigRequestHeaderSize) {
      status_ = inspect_extended_header();
    }
    if (status_ == RequestFrameStatus::NeedMore &&
        request_.bytes.size() == expected_size_) {
      status_ = RequestFrameStatus::Complete;
    }
  }
  return {status_, consumed};
}

RequestFrameStatus RequestFramer::eof() const noexcept {
  if (status_ == RequestFrameStatus::NeedMore && !request_.bytes.empty()) {
    return RequestFrameStatus::TruncatedInput;
  }
  return status_;
}

void RequestFramer::reset() {
  expected_size_ = kCoreRequestHeaderSize;
  status_ = RequestFrameStatus::NeedMore;
  request_ = {};
  request_.bytes.reserve(kCoreRequestHeaderSize);
}

void RequestFramer::enable_big_requests(
    const std::uint32_t maximum_length_units) {
  if (maximum_length_units < 2 ||
      maximum_length_units > kMaximumBigRequestLengthUnits)
    throw std::invalid_argument("invalid BIG-REQUESTS length limit");
  big_requests_enabled_ = true;
  maximum_big_length_units_ = maximum_length_units;
}

RequestFrameStatus RequestFramer::inspect_header() noexcept {
  request_.opcode = request_.bytes[0];
  request_.data = request_.bytes[1];
  ByteReader reader(std::span<const std::uint8_t>(request_.bytes).subspan(2),
                    order_);
  std::uint16_t ordinary_length_units{};
  if (!reader.read_u16(ordinary_length_units)) {
    return RequestFrameStatus::TruncatedInput;
  }
  request_.length_units = ordinary_length_units;
  if (request_.length_units == 0) {
    if (!big_requests_enabled_) return RequestFrameStatus::ZeroLength;
    request_.header_size = kBigRequestHeaderSize;
    expected_size_ = kBigRequestHeaderSize;
    request_.bytes.reserve(kBigRequestHeaderSize);
    return RequestFrameStatus::NeedMore;
  }
  if (request_.length_units > maximum_length_units_) {
    return RequestFrameStatus::TooLarge;
  }
  const auto size = core::checked_multiply(
      static_cast<std::size_t>(request_.length_units), std::size_t{4});
  if (!size || *size < kCoreRequestHeaderSize) {
    return RequestFrameStatus::TooLarge;
  }
  expected_size_ = *size;
  request_.bytes.reserve(expected_size_);
  return RequestFrameStatus::NeedMore;
}

RequestFrameStatus RequestFramer::inspect_extended_header() noexcept {
  ByteReader reader(std::span<const std::uint8_t>(request_.bytes).subspan(4),
                    order_);
  if (!reader.read_u32(request_.length_units))
    return RequestFrameStatus::TruncatedInput;
  if (request_.length_units < 2 ||
      request_.length_units > maximum_big_length_units_)
    return RequestFrameStatus::TooLarge;
  const auto size = core::checked_multiply(
      static_cast<std::size_t>(request_.length_units), std::size_t{4});
  if (!size || *size < kBigRequestHeaderSize)
    return RequestFrameStatus::TooLarge;
  expected_size_ = *size;
  request_.bytes.reserve(expected_size_);
  return RequestFrameStatus::NeedMore;
}

} // namespace gw::protocol::x11

File: crates/glasswyrm-x11/src/request.rs
use crate::ByteOrder;

pub const CORE_REQUEST_HEADER_SIZE: usize = 4;
pub const MAXIMUM_CORE_REQUEST_LENGTH_UNITS: u16 = u16::MAX;
pub const MAXIMUM_CORE_REQUEST_SIZE: usize = MAXIMUM_CORE_REQUEST_LENGTH_UNITS as usize * 4;
pub const BIG_REQUEST_HEADER_SIZE: usize = 8;
pub const MAXIMUM_BIG_REQUEST_SIZE: usize = 16 * 1024 * 1024;
pub const MAXIMUM_BIG_REQUEST_LENGTH_UNITS: u32 = (MAXIMUM_BIG_REQUEST_SIZE / 4) as u32;

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct FramedRequest {
    pub opcode: u8,
    pub data: u8,
    pub length_units: u32,
    pub header_size: usize,
    pub bytes: Vec<u8>,
}

impl Default for FramedRequest {
    fn default() -> Self {
        Self {
            opcode: 0,
            data: 0,
            length_units: 0,
            header_size: CORE_REQUEST_HEADER_SIZE,
            bytes: Vec::new(),
        }
    }
}

impl FramedRequest {
    #[must_use]
    pub fn body(&self) -> &[u8] {
        &self.bytes[self.header_size..]
    }

    #[must_use]
    pub fn core_size(&self) -> usize {
        self.bytes.len() - self.header_size + CORE_REQUEST_HEADER_SIZE
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum RequestFrameStatus {
    NeedMore,
    Complete,
    ZeroLength,
    TooLarge,
    TruncatedInput,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct RequestFrameResult {
    pub status: RequestFrameStatus,
    pub consumed: usize,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum RequestFramerConfigError {
    ZeroOrdinaryLimit,
    InvalidBigRequestsLimit,
}

#[derive(Debug)]
pub struct RequestFramer {
    order: ByteOrder,
    maximum_length_units: u16,
    maximum_big_length_units: u32,
    big_requests_enabled: bool,
    expected_size: usize,
    status: RequestFrameStatus,
    request: FramedRequest,
}

impl RequestFramer {
    pub fn new(
        order: ByteOrder,
        maximum_length_units: u16,
    ) -> Result<Self, RequestFramerConfigError> {
        if maximum_length_units == 0 {
            return Err(RequestFramerConfigError::ZeroOrdinaryLimit);
        }
        let mut request = FramedRequest::default();
        request.bytes.reserve(CORE_REQUEST_HEADER_SIZE);
        Ok(Self {
            order,
            maximum_length_units,
            maximum_big_length_units: MAXIMUM_BIG_REQUEST_LENGTH_UNITS,
            big_requests_enabled: false,
            expected_size: CORE_REQUEST_HEADER_SIZE,
            status: RequestFrameStatus::NeedMore,
            request,
        })
    }

    pub fn with_default_limit(order: ByteOrder) -> Self {
        Self::new(order, MAXIMUM_CORE_REQUEST_LENGTH_UNITS)
            .expect("the protocol maximum is nonzero")
    }

    pub fn consume(&mut self, input: &[u8]) -> RequestFrameResult {
        if self.status != RequestFrameStatus::NeedMore {
            return RequestFrameResult {
                status: self.status,
                consumed: 0,
            };
        }

        let mut consumed = 0;
        while consumed < input.len() && self.status == RequestFrameStatus::NeedMore {
            let needed = self.expected_size - self.request.bytes.len();
            let copy_size = needed.min(input.len() - consumed);
            self.request
                .bytes
                .extend_from_slice(&input[consumed..consumed + copy_size]);
            consumed += copy_size;

            if self.request.bytes.len() == CORE_REQUEST_HEADER_SIZE
                && self.expected_size == CORE_REQUEST_HEADER_SIZE
            {
                self.status = self.inspect_header();
            }
            if self.status == RequestFrameStatus::NeedMore
                && self.request.header_size == BIG_REQUEST_HEADER_SIZE
                && self.request.bytes.len() == BIG_REQUEST_HEADER_SIZE
                && self.expected_size == BIG_REQUEST_HEADER_SIZE
            {
                self.status = self.inspect_extended_header();
            }
            if self.status == RequestFrameStatus::NeedMore
                && self.request.bytes.len() == self.expected_size
            {
                self.status = RequestFrameStatus::Complete;
            }
        }
        RequestFrameResult {
            status: self.status,
            consumed,
        }
    }

    #[must_use]
    pub const fn eof(&self) -> RequestFrameStatus {
        if matches!(self.status, RequestFrameStatus::NeedMore) && !self.request.bytes.is_empty() {
            RequestFrameStatus::TruncatedInput
        } else {
            self.status
        }
    }

    #[must_use]
    pub const fn request(&self) -> &FramedRequest {
        &self.request
    }

    #[must_use]
    pub const fn expected_size(&self) -> usize {
        self.expected_size
    }

    pub fn enable_big_requests(
        &mut self,
        maximum_length_units: u32,
    ) -> Result<(), RequestFramerConfigError> {
        if !(2..=MAXIMUM_BIG_REQUEST_LENGTH_UNITS).contains(&maximum_length_units) {
            return Err(RequestFramerConfigError::InvalidBigRequestsLimit);
        }
        self.big_requests_enabled = true;
        self.maximum_big_length_units = maximum_length_units;
        Ok(())
    }

    pub fn enable_big_requests_with_default_limit(&mut self) {
        self.big_requests_enabled = true;
        self.maximum_big_length_units = MAXIMUM_BIG_REQUEST_LENGTH_UNITS;
    }

    #[must_use]
    pub const fn big_requests_enabled(&self) -> bool {
        self.big_requests_enabled
    }

    pub fn reset(&mut self) {
        self.expected_size = CORE_REQUEST_HEADER_SIZE;
        self.status = RequestFrameStatus::NeedMore;
        self.request = FramedRequest::default();
        self.request.bytes.reserve(CORE_REQUEST_HEADER_SIZE);
    }

    fn inspect_header(&mut self) -> RequestFrameStatus {
        self.request.opcode = self.request.bytes[0];
        self.request.data = self.request.bytes[1];
        self.request.length_units = u32::from(
            self.order
                .read_u16([self.request.bytes[2], self.request.bytes[3]]),
        );
        if self.request.length_units == 0 {
            if !self.big_requests_enabled {
                return RequestFrameStatus::ZeroLength;
            }
            self.request.header_size = BIG_REQUEST_HEADER_SIZE;
            self.expected_size = BIG_REQUEST_HEADER_SIZE;
            self.request.bytes.reserve(BIG_REQUEST_HEADER_SIZE);
            return RequestFrameStatus::NeedMore;
        }
        if self.request.length_units > u32::from(self.maximum_length_units) {
            return RequestFrameStatus::TooLarge;
        }
        let Some(size) = usize::try_from(self.request.length_units)
            .ok()
            .and_then(|units| units.checked_mul(4))
        else {
            return RequestFrameStatus::TooLarge;
        };
        if size < CORE_REQUEST_HEADER_SIZE {
            return RequestFrameStatus::TooLarge;
        }
        self.expected_size = size;
        self.request.bytes.reserve(size - self.request.bytes.len());
        RequestFrameStatus::NeedMore
    }

    fn inspect_extended_header(&mut self) -> RequestFrameStatus {
        self.request.length_units = self.order.read_u32([
            self.request.bytes[4],
            self.request.bytes[5],
            self.request.bytes[6],
            self.request.bytes[7],
        ]);
        if self.request.length_units < 2
            || self.request.length_units > self.maximum_big_length_units
        {
            return RequestFrameStatus::TooLarge;
        }
        let Some(size) = usize::try_from(self.request.length_units)
            .ok()
            .and_then(|units| units.checked_mul(4))
        else {
            return RequestFrameStatus::TooLarge;
        };
        if size < BIG_REQUEST_HEADER_SIZE {
            return RequestFrameStatus::TooLarge;
        }
        self.expected_size = size;
        self.request.bytes.reserve(size - self.request.bytes.len());
        RequestFrameStatus::NeedMore
    }
}
